components/scream/src/share/io/scream_scorpio_interface_iso_c2f.F90

```diff
@@ -256,7 +271,24 @@ subroutine grid_write_data_array_c2f_real(filename_in,varname_in,var_data_ptr) b
 
   end subroutine grid_write_data_array_c2f_real
 !=====================================================================!
-  subroutine grid_read_data_array_c2f(filename_in,varname_in,time_index,var_data_ptr) bind(c)
+  subroutine grid_read_data_array_c2f_int(filename_in,varname_in,time_index,var_data_ptr) bind(c)
```

Comment:
Perhaps, we could pass an integer `dtype`, as extra security. The user might pass an `int*`, thinking the data to be read has data type PIO_INT, but if for whatever reason it has data type PIO_REAL, it would cause the real fcn to be called, with bad memory errors coming right up; and even worse if it's the other way around, with the actual dtype fitted in the 'expected' dtype, so that no memory corruption happened, but the data would be completely wrong, leading to bugs hard to spot.

Note: this issue is already possible with the current code, since the F90 choses to cast the input c_ptr to int or real based on what the dtype of the stored var is, without any knowledge of what the actual dtype on the C side was.

Reply:
Actually, I recall that I tried this with just one function and it broke.  I'll have to go back and make it one function again and see what the error message was, but I remember that the fix was to break it into two different functions. 

From memory I believe there was a seg-fault the moment the data was actually accessed.  I do remember that I could see the data correctly inside the F90 code that this routine calls, so that worked fine.  But past that it no longer worked.

Reply:
And for sure I could have just messed something minor up.  Well worth exploring if it means getting cleaner code.